lisa/sut_orchestrator/azure/transformers.py

```diff
@@ -0,0 +1,259 @@
+    private_key_file: str = field(default="", metadata=schema.metadata(required=True))
+
+    # values for exported vhd. storage_account_name is optional, because it can
+    # be the default storage of LISA.
+    storage_account_name: str = ""
+    container_name: str = DEFAULT_VHD_CONTAINER_NAME
+    file_name_part: str = ""
+
+
+class VhdTransformer(Transformer):
+    """
+    convert an azure VM to VHD, which is ready to deploy.
+    """
+
+    __url_name = "url"
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "azure_vhd"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return VhdTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return [self.__url_name]
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: VhdTransformerSchema = self.runbook
+        platform_runbook_data = self._runbook_builder.partial_resolve(
+            constants.PLATFORM
+        )
+        platform_runbook = schema.Platform.schema().load(  # type: ignore
+            platform_runbook_data, many=True
+        )
+        platform = load_platform(platform_runbook)
+        assert isinstance(
+            platform, AzurePlatform
+        ), f"'{self.type_name}' support only Azure platform"
+
+        platform.initialize()
+        compute_client = get_compute_client(platform)
+        virtual_machine = compute_client.virtual_machines.get(
+            runbook.resource_group_name, runbook.vm_name
+        )
+
+        node = self._prepare_virtual_machine(platform, virtual_machine)
+
+        vhd_location = self._export_vhd(platform, virtual_machine)
+
+        self._recover_vm(platform, virtual_machine, node)
+
+        return {self.__url_name: vhd_location}
+
+    def _prepare_virtual_machine(
+        self, platform: AzurePlatform, virtual_machine: Any
+    ) -> RemoteNode:
+        runbook: VhdTransformerSchema = self.runbook
+        if not runbook.public_address:
+            runbook.public_address = self._get_public_ip_address(
+                platform, virtual_machine
+            )
+
+        node_runbook = schema.RemoteNode(
+            name=runbook.vm_name,
+            public_address=runbook.public_address,
+            port=runbook.public_port,
+            username=runbook.username,
+            password=runbook.password,
+            private_key_file=runbook.private_key_file,
+        )
+        node = RemoteNode(
+            runbook=node_runbook, index=0, logger_name=f"{self.type_name()}_vm"
+        )
+        node.features = Features(node, platform)
+        node_context = get_node_context(node)
+        node_context.vm_name = runbook.vm_name
+        node_context.resource_group_name = runbook.resource_group_name
+
+        node.set_connection_info_by_runbook()
+        node.initialize()
+
+        # prepare vm for exporting
+        wa = node.tools[Waagent]
+        wa.deprovision()
+        node.execute("export HISTSIZE=0", shell=True)
```

Comment:
have you tried this, seems after deprovision, the network of the VM is disconnected.

Reply:
It works well, the VM is not disconnected.

Reply:
Why that isolated HISTSIZE change? It will only work for that session.

Reply:
The logic is copied from V2. I think it's used to prevent future history after deprovisioning.

Reply:
Yeah, but it will only apply to children of that shell session, which dies right after the command. It will do absolutely nothing to future SSH calls, regarding that.

Reply:
After this command, the VM will be shutdown, and export a VHD. There is no other sessions before shutdown.

Reply:
Which is even a stronger signal that that export is useless there, no?

Reply:
For the usage of this command, it's copied from v2. It should be used to truncate existing history.